[Parse] Avoid skipping bodies with `#sourceLocation`

hamishknight opened the pull request:
`#sourceLocation` requires setting and restoring state in the parser, so we cannot skip any bodies containing `#sourceLocation` tokens. Member skipping was already doing this, extend it to function body skipping too.

hamishknight commented:
@swift-ci please test

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -7080,25 +7085,25 @@ bool Parser::canDelayMemberDeclParsing(bool &HasOperatorDeclarations,
                           HasPotentialRegexLiteral);
-  if (!HasPoundDirective && !HasPotentialRegexLiteral)
+  if (!HasPoundDirective && !HasPotentialRegexLiteral) {
+    ASSERT(!HasPoundSourceLocation);
```
Why can't `HasPoundSourceLocation` be true here?

Out of curiosity: Why are you using uppercase `ASSERT` here? Don’t we usually spell `assert` with lowercase?

ahoppen replied:
Ah, I see. Might be worth mentioning in a comment but not too important.

hamishknight committed:
[Parse] Allow skipping member bodies following `#sourceLocation`

If the bodies themselves don't have a
`#sourceLocation` in them, there should be no
harm in skipping; the parsed virtual files get
recorded on the SourceManager, so will be
accessible when doing delayed parsing. The
`InPoundLineEnvironment` flag will be inaccurate,
but that's only needed for the parsing of
`#sourceLocation` itself.

hamishknight committed:
[Parse] Avoid skipping bodies with `#sourceLocation`

`#sourceLocation` requires setting and restoring
state in the parser, so we cannot skip any bodies
containing `#sourceLocation` tokens. Member skipping
was already doing this, extend it to function
body skipping too.

hamishknight commented:
@swift-ci please smoke test
